SQL Exception when using Collection<String> as parameter with custom Query

### Expected Behavior

_No response_

### Actual Behaviour

It throws an SQL Exception instead of deleting the entries.
```java
22:21:25.754 [main] DEBUG i.m.t.j.DataSourceTransactionManager -
        Initiating transaction rollback on application exception
io.micronaut.data.exceptions.DataAccessException: Error executing SQL Query: ERROR: syntax error at or near "$3"
  Position: 156
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$findOne$3(DefaultJdbcRepositoryOperations.java:463)
	at io.micronaut.transaction.support.AbstractSynchronousTransactionManager.executeRead(AbstractSynchronousTransactionManager.java:155)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findOne(DefaultJdbcRepositoryOperations.java:406)
	at io.micronaut.data.runtime.intercept.DefaultFindOneInterceptor.intercept(DefaultFindOneInterceptor.java:45)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:114)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:88)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:138)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at club.devcord.devmarkt.repositories.RoleRepo$Intercepted.deletePermissions(Unknown Source)
	at club.devcord.devmarkt.auth.TestFactory.testLol(TestFactory.java:39)
	at club.devcord.devmarkt.auth.$TestFactory$TestLol0$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3281)
	at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2664)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1932)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:237)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3453)
	at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3883)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:329)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:183)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:72)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:313)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:299)
	at club.devcord.devmarkt.Application.main(Application.java:24)
Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "$3"
  Position: 156
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2674)
Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "$3"
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2364)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:354)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:484)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:404)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:162)
	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:114)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$findOne$3(DefaultJdbcRepositoryOperations.java:410)
	... 24 common frames omitted
	```

### Steps To Reproduce

I have following repository:
```java
@JdbcRepository
public interface RoleRepo extends CrudRepository<Role, Integer> {


  @io.micronaut.data.annotation.Query(value = """
WITH to_delete_role AS (
    SELECT id FROM roles WHERE name = :roleName
),
to_delete_perm AS (
     SELECT id FROM permissions WHERE operation = :operation AND query IN :queries
 )
DELETE FROM role_permissions WHERE role_id IN (SELECT id FROM to_delete_role) AND permission_id IN (SELECT id FROM to_delete_perm);
""")
  void deletePermissions(String roleName, Operation operation, Collection<String> queries);
  }
  ```
  
  The method call: 
  ```java
      repo.deletePermissions("test", Operation.MUTATION, Set.of("testMutation1", "testQuery2"));
  ```
  
As info the query runs without problems without micronaut.

### Environment Information

- Manjaro
- Java 17

### Example Application

_No response_

### Version

3.3.4

--- a/data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/JdbcQueryStatement.java
+++ b/data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/JdbcQueryStatement.java
@@ -151,7 +151,7 @@ public class JdbcQueryStatement implements QueryStatement<PreparedStatement, Int
                 statement.setArray(index, (Array) value);
             } else if (value != null) {
                 if (value.getClass().isEnum()) {
-                    statement.setString(index, ((Enum) value).name());
+                    statement.setObject(index, value, java.sql.Types.OTHER);
                 } else {
                     statement.setObject(index, value);
                 }

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/postgres/PostgresEnumSpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/postgres/PostgresEnumSpec.groovy
@@ -1,0 +1,133 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.jdbc.postgres
+
+import io.micronaut.context.ApplicationContext
+import io.micronaut.core.convert.ConversionContext
+import io.micronaut.data.annotation.GeneratedValue
+import io.micronaut.data.annotation.Id
+import io.micronaut.data.annotation.MappedEntity
+import io.micronaut.data.annotation.MappedProperty
+import io.micronaut.data.annotation.Query
+import io.micronaut.data.annotation.TypeDef
+import io.micronaut.data.jdbc.annotation.JdbcRepository
+import io.micronaut.data.jdbc.convert.JdbcConversionContext
+import io.micronaut.data.model.DataType
+import io.micronaut.data.model.query.builder.sql.Dialect
+import io.micronaut.data.model.runtime.convert.AttributeConverter
+import io.micronaut.data.repository.PageableRepository
+import jakarta.inject.Singleton
+import org.testcontainers.containers.JdbcDatabaseContainer
+import spock.lang.AutoCleanup
+import spock.lang.Shared
+import spock.lang.Specification
+
+import java.sql.Array
+import java.sql.Connection
+
+class PostgresEnumsSpec extends Specification implements PostgresTestPropertyProvider {
+
+    @AutoCleanup
+    @Shared
+    ApplicationContext context = ApplicationContext.run(properties)
+
+    @Override
+    void startContainer(JdbcDatabaseContainer container) {
+        PostgresTestPropertyProvider.super.startContainer(container)
+        try (Connection connection = container.createConnection("")) {
+            connection.prepareCall("""CREATE TYPE happiness AS ENUM ('happy', 'very_happy', 'ecstatic');""").execute()
+        }
+    }
+
+    def "test native enums"() {
+        given:
+            def repo = context.getBean(EnuRepo)
+            def e1 = repo.save(new EnuEntity(null, "ABC", Happiness.ecstatic))
+            def e2 = repo.save(new EnuEntity(null, "XYZ1", Happiness.happy))
+            def e3 = repo.save(new EnuEntity(null, "XYZ2", Happiness.happy))
+        when:
+            def e = repo.findById(e1.id).get()
+        then:
+            e.happiness == Happiness.ecstatic
+            repo.findByName("ABC").id == e.id
+            repo.findByHappiness(Happiness.ecstatic).id == e.id
+            repo.findByHappinessIn([Happiness.happy, Happiness.very_happy]).size() == 2
+            repo.findHappinessByHappinessIn([Happiness.happy, Happiness.very_happy]) == [Happiness.happy, Happiness.happy]
+            repo.findByHappinessAsArray(Happiness.happy, Happiness.very_happy).size() == 2
+            repo.findHappinessByHappinessAsArray(Happiness.happy, Happiness.very_happy) == [Happiness.happy, Happiness.happy]
+        when:
+            repo.update(e1.id, Happiness.very_happy)
+            e = repo.findById(e1.id).get()
+        then:
+            e.happiness == Happiness.very_happy
+    }
+
+}
+
+@MappedEntity("pg_enumz")
+class EnuEntity {
+    @Id
+    @GeneratedValue
+    Long id
+    String name
+    @MappedProperty(definition = "happiness", type = DataType.OBJECT)
+    Happiness happiness
+
+    EnuEntity(Long id, String name, Happiness happiness) {
+        this.id = id
+        this.name = name
+        this.happiness = happiness
+    }
+}
+
+enum Happiness {
+    happy, very_happy, ecstatic
+}
+
+@JdbcRepository(dialect = Dialect.POSTGRES)
+interface EnuRepo extends PageableRepository<EnuEntity, Long> {
+
+    EnuEntity findByName(String name)
+
+    EnuEntity findByHappiness(Happiness val)
+
+    void update(@Id Long id, Happiness happiness)
+
+    Collection<EnuEntity> findByHappinessIn(Collection<Happiness> vals)
+
+    Collection<Happiness> findHappinessByHappinessIn(Collection<Happiness> vals)
+
+    @Query("select * from pg_enumz where happiness = ANY (:vals)")
+    Collection<EnuEntity> findByHappinessAsArray(@TypeDef(type = DataType.OBJECT, converter = ArraySetter.class) Happiness... vals)
+
+    @Query("select happiness from pg_enumz where happiness = ANY (:vals)")
+    Collection<Happiness> findHappinessByHappinessAsArray(@TypeDef(type = DataType.OBJECT, converter = ArraySetter.class) Happiness... vals)
+
+}
+
+@Singleton
+class ArraySetter implements AttributeConverter<Object, Array> {
+
+    @Override
+    Array convertToPersistedValue(Object entityValue, ConversionContext context) {
+        return (context as JdbcConversionContext).connection.createArrayOf("happiness", entityValue)
+    }
+
+    @Override
+    Object convertToEntityValue(Array persistedValue, ConversionContext context) {
+        throw new IllegalStateException("Not supported!")
+    }
+}

--- a/data-processor/src/main/java/io/micronaut/data/processor/model/criteria/impl/SourceParameterExpressionImpl.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/model/criteria/impl/SourceParameterExpressionImpl.java
@@ -18,6 +18,7 @@ package io.micronaut.data.processor.model.criteria.impl;
 import io.micronaut.core.annotation.Internal;
 import io.micronaut.data.annotation.AutoPopulated;
 import io.micronaut.data.annotation.Expandable;
+import io.micronaut.data.annotation.TypeDef;
 import io.micronaut.data.model.Association;
 import io.micronaut.data.model.DataType;
 import io.micronaut.data.model.PersistentProperty;
@@ -89,6 +90,7 @@ public final class SourceParameterExpressionImpl extends ParameterExpressionImpl
             Objects.requireNonNull(parameterElement);
             int index = Arrays.asList(parameters).indexOf(parameterElement);
             DataType dataType = getDataType(null, parameterElement);
+            String converter = parameterElement.stringValue(TypeDef.class, "converter").orElse(null);
             boolean isExpandable = isExpandable(bindingContext, dataType);
             return new QueryParameterBinding() {
                 @Override
@@ -104,6 +106,11 @@ public final class SourceParameterExpressionImpl extends ParameterExpressionImpl
                 @Override
                 public DataType getDataType() {
                     return dataType;
+                }
+
+                @Override
+                public String getConverterClassName() {
+                    return converter;
                 }
 
                 @Override

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/mapper/QueryStatement.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/mapper/QueryStatement.java
@@ -64,11 +64,15 @@ public interface QueryStatement<PS, IDX> {
         switch (dataType) {
             case STRING:
             case JSON:
+                String str;
                 if (value instanceof CharSequence) {
-                    return setString(statement, index, value.toString());
+                    str = value.toString();
+                } else if (value instanceof Enum) {
+                    str = ((Enum<?>) value).name();
                 } else {
-                    return setString(statement, index, convertRequired(value, String.class));
+                    str = convertRequired(value, String.class);
                 }
+                return setString(statement, index, str);
             case INTEGER:
                 if (value instanceof Number) {
                     return setInt(statement, index, ((Number) value).intValue());

